### PR TITLE
fix: support `Node16` and `NodeNext` module resolution in `experimentalDts`

### LINKED ISSUE #1023: TypeScript files are seen as JavaScript when tsup resolves types of an external package in certain conditions
If a package points a subpath export (`exports: { "./a": "./a.ts" }` in package.json) to a TypeScript file (“a.ts”: `import {…} from "./b.ts"`), and you use that package as a dependency, when you run tsup with `dts: { resolve: true }`:
the file (“a.ts”) itself is treated as TypeScript code, but TypeScript files (like “b.ts”) imported by the former file are treated as plain JavaScript codes.

If “a.ts” is set as the main export (`exports: { ".": "./ts" }`), everything works.

## Reproduction

https://github.com/umajho/tsup-reproduction-dts-resolve-pnpm-workspace

in “packages/support-lib”:
``` ts
// foo.ts:
export type Foo = "foo";
export const foo: Foo = "foo";

// bar-that-imports-foo.ts:
export * from "./foo";

// package.json:
{
  // …
  "exports": {
    "./foo": "./foo.ts",
    "./bar-that-imports-foo": "./bar-that-imports-foo.ts",
    ".": "./bar-that-imports-foo.ts"
  }
}
```

---

`cd packages/lib-to-be-built` 

### `pnpm run build-foo-dts-resolve` ⭕️

Importing from “support-lib” with the subpath “./foo”, and building the library using tsup with `dts: { resolve: true }`:

``` ts
// lib-imports-foo.ts:
import { Foo, foo } from "@tsup-reproduction-dts-resolve-pnpm-workspace/support-lib/foo";
export const baz: Foo = foo;
```

<details>
<summary>success</summary>

```
➜  lib-to-be-built git:(main) pnpm run build-foo-dts-resolve

> @tsup-reproduction-dts-resolve-pnpm-workspace/lib-to-be-built@1.0.0 build-foo-dts-resolve /private/tmp/test/packages/lib-to-be-built
> tsup --config tsup.foo-dts-resolve.config.ts

CLI Building entry: lib-imports-foo.ts
CLI Using tsconfig: tsconfig.json
CLI tsup v7.2.0
CLI Using tsup config: /private/tmp/test/packages/lib-to-be-built/tsup.foo-dts-resolve.config.ts
CLI Target: es2020
CLI Cleaning output folder
ESM Build start
ESM dist/lib-imports-foo.js 98.00 B
ESM ⚡️ Build success in 10ms
DTS Build start
DTS ⚡️ Build success in 474ms
DTS dist/lib-imports-foo.d.ts 60.00 B
```
</details>

### `pnpm run build-bar-using-main-dts-resolve` ⭕️

Importing from “support-lib” with the main path, and building the library using tsup with `dts: { resolve: true }`:

``` ts
// lib-imports-bar-using-main.ts:
import { Foo, foo } from "@tsup-reproduction-dts-resolve-pnpm-workspace/support-lib";
export const baz: Foo = foo;
```

<details>
<summary>success</summary>

```
➜  lib-to-be-built git:(main) pnpm run build-bar-using-main-dts-resolve

> @tsup-reproduction-dts-resolve-pnpm-workspace/lib-to-be-built@1.0.0 build-bar-using-main-dts-resolve /private/tmp/test/packages/lib-to-be-built
> tsup --config tsup.bar-using-main-dts-resolve.config.ts

CLI Building entry: lib-imports-bar-using-main.ts
CLI Using tsconfig: tsconfig.json
CLI tsup v7.2.0
CLI Using tsup config: /private/tmp/test/packages/lib-to-be-built/tsup.bar-using-main-dts-resolve.config.ts
CLI Target: es2020
CLI Cleaning output folder
ESM Build start
ESM dist/lib-imports-bar-using-main.js 109.00 B
ESM ⚡️ Build success in 12ms
DTS Build start
DTS ⚡️ Build success in 453ms
DTS dist/lib-imports-bar-using-main.d.ts 60.00 B
```

</details>

### `pnpm run build-bar-dts-resolve` ❌

Importing from “support-lib” with the subpath “./bar-that-imports-foo”, and building the library using tsup with `dts: { resolve: true }`:

``` ts
// lib-imports-bar.ts:
import { Foo, foo } from "@tsup-reproduction-dts-resolve-pnpm-workspace/support-lib/bar-that-imports-foo";
export const baz: Foo = foo;
```

<details open>
<summary>failed</summary>

```
➜  lib-to-be-built git:(main) pnpm run build-bar-dts-resolve

> @tsup-reproduction-dts-resolve-pnpm-workspace/lib-to-be-built@1.0.0 build-bar-dts-resolve /private/tmp/test/packages/lib-to-be-built
> tsup --config tsup.bar-dts-resolve.config.ts

CLI Building entry: lib-imports-bar.ts
CLI Using tsconfig: tsconfig.json
CLI tsup v7.2.0
CLI Using tsup config: /private/tmp/test/packages/lib-to-be-built/tsup.bar-dts-resolve.config.ts
CLI Target: es2020
CLI Cleaning output folder
ESM Build start
ESM dist/lib-imports-bar.js 98.00 B
ESM ⚡️ Build success in 16ms
DTS Build start
Error parsing: /private/tmp/test/packages/support-lib/foo.ts:1:7
Error: error occured in dts build
    at Worker.<anonymous> (/private/tmp/test/node_modules/.pnpm/tsup@7.2.0_typescript@5.2.2/node_modules/tsup/dist/index.js:2294:26)
    at Worker.emit (node:events:511:28)
    at MessagePort.<anonymous> (node:internal/worker:263:53)
    at [nodejs.internal.kHybridDispatch] (node:internal/event_target:762:20)
    at exports.emitMessage (node:internal/per_context/messageport:23:28)
DTS Build error
Unexpected token (Note that you need plugins to import files that are not JavaScript)
1: export type Foo = "foo";
          ^
2: 
3: export const foo: Foo = "foo";
 ELIFECYCLE  Command failed with exit code 1.
```

</details>

## Upvote & Fund

- We're using [Polar.sh](https://polar.sh/egoist) so you can upvote and help fund this issue.
- We receive the funding once the issue is completed & confirmed by you.
- Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/egoist/tsup/issues/1023">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/egoist/tsup/issues/1023/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### LINKED ISSUE #1224: `experimentalDts` does not work with `Node16` and `NodeNext` module resolution
## **Overview**

Currently `experimentalDts` does not emit the correct type definitions when `moduleResolution` is set to either [**`Node16`** or **`NodeNext`**](https://www.typescriptlang.org/docs/handbook/modules/reference.html#node16-nodenext-1). The `_tsup-dts-rollup.d.ts` file is trying to import from the `index.d.ts` file which causes a circular reference.

<details><summary><h3><b><code>_tsup-dts-rollup.d.ts</code></b></h3></summary>

rollup

</details>

<details><summary><h3><b><code>index.d.ts</code></b></h3></summary>

index-circular

</details>

## Reproduction

1. Clone the branch containing the reproduction by running:

```bash
git clone -b experimentalDts-NodeNext https://github.com/aryaemami59/repros
```

2. Install the dependencies:

```bash
yarn install
```

3. Run the following command:

```bash
yarn run build:node-next
```

4. Inspect the output in the `dist` directory.

---

I will submit a PR for the fix.

### PR DESCRIPTION
**Pull request by aryaemami59**
## **This PR**:

- [X] Fixes `experimentalDts` to work correctly under TypeScript's [**`Node16`** and **`NodeNext`**](https://www.typescriptlang.org/docs/handbook/modules/reference.html#node16-nodenext-1) module resolution.
- [X] Resolves #1224.
- [X] Resolves #1023.

### CODE REVIEW

File: src/utils.ts [hunk 1]
```diff
@@ -242,3 +242,32 @@ export function writeFileSync(filePath: string, content: string) {
...
+ *
+ * @internal
+ */
+export function replaceDtsWithJsExtensions(dtsFilePath: string) {
+  return dtsFilePath.replace(
+    /\.d\.(ts|mts|cts)$/,
+    (_, fileExtension: string) => {
+      switch (fileExtension) {
+        case 'ts':
+        case 'tsx':
```
Comment (ocavue) [inline src/utils.ts:262]: The regex above doesn't match `tsx`, so either we should remove `tsx` here, or add `tsx` in the regex above.

Comment (aryaemami59) [reply src/utils.ts:262]: sure!

Comment (ocavue) [reply src/utils.ts:262]: I did a test and `tsc` will compile `component.tsx` into `component.d.ts`, so it seems that `.d.tsx` isn't a thing. We should simply remove `case 'tsx'`.

Comment (aryaemami59) [reply src/utils.ts:262]: Sounds good!

**Review by ocavue — CHANGES_REQUESTED** [review]
To matches the behavior of `tsc`'s output, `.js` should only be added for `Node16`/`NodeNext`, not for `Bundler` etc.

**Comment by aryaemami59** [comment]
@ocavue `tsc`'s `Bundler` module resolution can still handle (and output) relative `.js\.mjs\.cjs` file extensions.

**Comment by ocavue** [comment]
> `tsc`'s Bundler module resolution can still handle (and output) relative `.js\.mjs\.cjs` file extensions.

Thanks. TIL. 

Does other `moduleResolution` options like `Node` can handle relative file extensions too?

**Comment by aryaemami59** [comment]
@ocavue I think so, but just to be safe let me double check.

**Comment by aryaemami59** [comment]
> Does other `moduleResolution` options like `Node` can handle relative file extensions too?

@ocavue Answer is yes!

**Comment by ocavue** [comment]
I still believe it’s safer to match `tsc`’s behavior (i.e., whether to include the `.js `extension in the generated `.d.ts` files). The output of `tsc` is the de facto standard, and there may be other tools besides TypeScript that consume `.d.ts` files.

In my opinion, a better approach would be to use `moduleResolution` to determine whether to add the `.js` extension.

**Comment by aryaemami59** [comment]
@ocavue Can't say I agree with that but if we wanted to offer some more flexibility we could put this functionality behind an option? Something like `addRelativePathExtension`? The issue is as it stands right now (prior to this PR) the resulting type definitions for both `ESM` and `CJS` end up importing from the same `_tsup-dts-rollup` file.

**Comment by ocavue** [comment]
`addRelativePathExtension` seems too complex to me.

**Comment by aryaemami59** [comment]
@ocavue What do you think of removing `_tsup-dts-rollup` from the final `outDir`?

> `addRelativePathExtension` seems too complex to me.

I was going for something that was explicit and self-explanatory. While it may sound complicated, it sounds fairly obvious what it does. But fair enough, what would you have in mind?

**Comment by ocavue** [comment]
> What do you think of removing _tsup-dts-rollup files from the final outDir?

Most packages only have one entrypoint, and removing `_tsup-dts-rollup` is a good idea for them. `_tsup-dts-rollup.d.ts` is a silly workaround for the fact that `@microsoft/api-extractor` doesn't support multi-entrypoints packages. Maybe a better solution is only output `_tsup-dts-rollup.d.ts` if there are more than one entrypoints.

---
 

I just don't think it's necessary to config `addRelativePathExtension` manually. Other `.d.ts` generator (i.e. `tsc`) don't have such an option.

**Comment by ocavue** [comment]
> I can try to see if it would be feasible to remove _tsup-dts-rollup altogether.

Thanks. It's better to be a separate PR by the way

**Comment by aryaemami59** [comment]
@ocavue

> Thanks. It's better to be a separate PR by the way

Sounds good, so one PR to remove `_tsup-dts-rollup` and one PR (whicch is this one) to get `experimentalDts` working in `NodeNext` and `Node16`, correct?

**Comment by ocavue** [comment]
> After all the reason why we use experimentalDts or dts in the first place is so we don't have to use tsc

It's not important which tool we use to generate these .d.ts files, it's important that whoever consumes these .d.ts files won't get compatibility issue, and the safest option is matching tsc's output.

Not related: experimentalDts still needs tsc to output .d.ts files.

**Comment by ocavue** [comment]
> Sounds good, so one PR to remove `_tsup-dts-rollup` and one PR (whicch is this one) to get `experimentalDts` working in `NodeNext` and `Node16`, correct?

yes

**Comment by aryaemami59** [comment]
@ocavue I know you wanted separate PRs, but it seems like by removing `_tsup-dts-rollup`, we can take care of this problem too. 2 birds, 1 stone. Thoughts?

**Review by ocavue — CHANGES_REQUESTED** [review]
The removal of `_tsup-dts-rollup` breaks multiple-entrypoints packages.

Let's say I have a package called `my-lib`, which has the following files

```
my-lib/
  src/
    foo.ts
    bar.ts
    internal.ts
  package.json
```

In `internal.ts`, I define a class `export class SharedClass { ... }`, and in both `foo.ts` and `bar.ts`, I re-export this `SharedClass`.

In the generated `.d.ts` files, there should be only one `export declare class Shared`, which is currently placed into `dist/_tsup-dts-rollup.d.ts`. 

If you remove `dist/_tsup-dts-rollup.d.ts`, and having two `export declare class SharedClass` in `dist/foo.d.ts` and `dist/bar.d.ts`, you could get the following error:

```ts
import { type SharedClass as Bar } from 'my-lib/bar'
import { SharedClass as Foo } from 'my-lib/foo'

const foo: Foo = new Foo()
const bar: Bar = foo
   // ^: Type 'import("/private/tmp/test/packages/my-lib/dist/foo").SharedClass' is not assignable to type 'import("/private/tmp/test/packages/my-lib/dist/bar").SharedClass'.

console.log(bar, foo)

```

**Comment by aryaemami59** [comment]
@ocavue I tried recreating the issue that you said this would cause, but I can't. I'm going to try some more, but is there anyway you could provide a more concrete example so I can make sure this PR doesn't cause the issues you're describing?

**Comment by ocavue** [comment]
>  is there anyway you could provide a more concrete example

Here you are:

1. Open https://stackblitz.com/github/issueset/tsup-multi-dts or clone https://github.com/issueset/tsup-multi-dts
2. Run `pnpm install && pnpm run build`
3. You will get the following error 

```

src/bad.ts:5:7 - error TS2322: Type 'import("/private/tmp/tsup-multi-dts/packages/my-lib-bad/dist/foo").SharedClass' is not assignable to type 'import("/private/tmp/tsup-multi-dts/packages/my-lib-bad/dist/bar").SharedClass'.
  Types of property 'symbol' are incompatible.
    Type 'typeof import("/private/tmp/tsup-multi-dts/packages/my-lib-bad/dist/foo").sharedSymbol' is not assignable to type 'typeof import("/private/tmp/tsup-multi-dts/packages/my-lib-bad/dist/bar").sharedSymbol'.
```

**Comment by aryaemami59** [comment]
@ocavue ahh I see, this is mostly due to how TypeScript handles [`unique symbol` types](https://www.typescriptlang.org/docs/handbook/symbols.html#unique-symbol) as a matter of fact this can be a bit of an issue with `dts` as well, not just `experimentalDts`, I forked your repro and added an example for this on another branch you're more than welcome to look at it if you want, you can clone it by running:

```bash
git clone -b tsup-dts-unique-symbols https://github.com/aryaemami59/tsup-multi-dts
```

**Comment by ocavue** [comment]
> this can be a bit of an issue with `dts` as well

`rollup-plugin-dts` works well with the following `tsup.config.ts`.

```ts
import { defineConfig } from "tsup";

export default defineConfig([
  {
    entry: ["src/foo.ts", "src/bar.ts"],
    format: ["esm"],
    dts: true,
  },
]);
```

---

Anyway, as I said before, it might be better to write a separate PR for the removal of `_tsup-dts-rollup`. It might not be as easy as it looks. Meanwhile, supporting `Node16` and `NodeNext` is relatively simple and straightforward.

**Comment by aryaemami59** [comment]
@ocavue That's fair, I'll adjust this PR accordingly.

**Comment by aryaemami59** [comment]
@ocavue Ok this should be good to go, I tried to do what you suggested in [this comment](https://github.com/egoist/tsup/pull/1225#issuecomment-2413389622) but decided not to do it, here are the reasons: 
1. It introduces a number of issues, especially when `format` is set to `['esm', 'cjs']`. 
2. TypeScript explicitly expects `js` extensions for relative imports, as detailed in [their documentation](https://www.typescriptlang.org/docs/handbook/modules/theory.html#module-specifiers-are-not-transformed).
3. The same approach is used by `rollup-plugin-dts` in `options.dts`.
4. When outputting both ESM and CJS type definitions, what we’re really doing is running `api-extractor` twice. Once normally and once as if all files have `.cjs` or `.mjs` extensions, depending on the `type` field in `package.json`.
5. All other module resolutions can handle explicit extensions.

Please let me know if there is anything else you want me to do.

**Review by ocavue — APPROVED** [review]
Great! I love your detailed comment. I'll merge it and release.

**Pull request merged**